### PR TITLE
#159987404 users should be able to create their profiles

### PR DESCRIPTION
**Pull request by joeeasy**
**What does this PR do?**
Implements a feature that allows users to: 
- view other users profile

- edit their profile

- upload a profile picture

- add a bio

**Description of Task to be completed?**
- install `datauri`
- install `Multer`
- install `Cloudinary`
- write tests for the endpoint
- write a validator middleware

**How should this be manually tested?**
Clone the repository.
Checkout the `ft-users-can-update-profile` branch.
Run `NPM START` on your local machine.
Run database migrations using `npm run migrate`.
Start the server using `NPM START`.

**You can test the following endpoints using postman**
PUT `localhost:8000/api/v1/users/update` to update a user's profile
GET `localhost:8000/api/v1/users/:userId` to view a user's profile
HEADERS
- authorization -> paste your token there
- Content-Type -> application/form-data

REQUIRED FIELDS
-    username: `string`
-   email: `string`

**What are the relevant pivotal tracker stories?**
#159987404

### CODE REVIEW

File: server/controllers/userController.js [hunk 1]
```diff
@@ -66,6 +70,76 @@ const userController = {
...
+   * @description Allows users to edit their profile
+   * @param {Object} req The request object
+   * @param {Object} res The the response object
+   * @returns {Object} The modified profile details
+   */
+  updateProfile: (req, res) => {
+    multerUploads(req, res, () => {
+      const { currentUser } = req;
+      let image = '';
+      const file = !req.file ? '../../upload/dummy-profile.png' : dataUri(req);
```
Comment (iAmao) [inline server/controllers/userController.js]: I'm guessing, you are uploading a default image if the user did not pass an image. Instead, what we should do is have a default image URL in the column, and not make any call to cloudinary if the user did not upload a file.

**Review by iAmao — CHANGES_REQUESTED** [review]

File: swagger.json [hunk 2]
```diff
@@ -331,7 +331,134 @@
           }
         }
       }
-    },
+
```
Comment (jojitoon) [inline swagger.json]: remove spaces

**Review by jojitoon — APPROVED** [review]

File: test/users.test.spec.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        image: './uploads/dummy-profile.png',
+      })
+      .end((err, res) => {
+        expect(res).to.have.status(401);
+        expect(res.body).to.be.an('object');
+        expect(res.body.data.message).to.equal('Failed to authenticate token! Valid token required');
+        done();
+      });
+  });
+  it("should return successfully when a user's profile is successfully updated", (done) => {
```
Comment (iAmao) [inline test/users.test.spec.js]: Prefer single quotes over double quotes as suggested by the style guide

File: test/users.test.spec.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+      .request(app)
+      .put('/api/v1/users/update')
+      .set('Authorization', verifiedToken)
+      .send({
+        firstname: 'joe',
+        lastname: 'easy',
+        email: 'joe.test@testmail.com',
+        username: 'joeeasy',
+        bio: 'Hello, here is a brief information about me',
+        interest: 'gaming, movies, music, tech',
```
Comment (iAmao) [inline test/users.test.spec.js:92]: in future we might want to have this as a table

File: test/users.test.spec.js [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+      })
+      .end((err, res) => {
+        expect(res).to.have.status(200);
+        expect(res.body).to.be.an('object');
+        expect(res.body.data.message).to.equal('Your profile has been updated successfully');
```
Comment (iAmao) [inline test/users.test.spec.js]: You are not asserting that any value actually changed here

File: test/users.test.spec.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        done();
+      });
+  });
+  it("should return successfully when a user's profile is successfully updated", (done) => {
```
Comment (iAmao) [inline test/users.test.spec.js]: Prefer single quotes over double quotes as suggested by the style guide

File: test/users.test.spec.js [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+      .request(app)
+      .put('/api/v1/users/update')
+      .set('Authorization', verifiedToken)
+      .send({
+        firstname: 'joe',
+        lastname: 'easy',
+        email: 'joe.test@testmail.com',
+        username: 'joeeasy',
+        bio: 'Hello, here is a brief information about me',
+        // interest: 'gaming, movies, music, tech',
```
Comment (iAmao) [inline test/users.test.spec.js]: avoid pushing commented code

File: test/users.test.spec.js (hunk 6 above)
Comment (iAmao) [reply test/users.test.spec.js]: Also, isn't this a duplicate test?

File: test/users.test.spec.js [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+      .field('username', 'joeeasy')
+      .field('firstname', 'jehonadab')
+      .field('lastname', 'hehehehehe')
+      .field('bio', 'Here is a brief information about me')
+      .field('interest', 'runing')
+      .field('email', 'joeeasy@gmail.com')
+      .type('form')
+      .end((err, res) => {
+        expect(res).to.have.status(200);
+        expect(res.body.data.message).to.equal('Your profile has been updated successfully');
```
Comment (iAmao) [inline test/users.test.spec.js]: There is actually no response validating that this works

File: test/users.test.spec.js [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+});
+
+describe('Checks the validity of the user', () => {
+  it('should return a user', (done) => {
+    chai.request(app)
+      .get('/api/v1/users/1')
+      .set('Authorization', verifiedToken)
+      .end((error, res) => {
+        expect(res).to.have.status(200);
+        assert.isObject(res.body, 'is an object containing the user details');
```
Comment (iAmao) [inline test/users.test.spec.js]: expect the user's details as well

File: test/users.test.spec.js [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+        expect(res.body.data.message).to.equal('No user found');
+        done();
+      });
+  });
+  it('should not return  a user if id is an alphabet', (done) => {
+    chai.request(app)
+      .get('/api/v1/users/abcd')
+      .set('Authorization', verifiedToken)
+      .end((message, res) => {
+        expect(res).to.have.status(400);
```
Comment (iAmao) [inline test/users.test.spec.js]: Server should return 404 In this case

File: test/users.test.spec.js [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+        expect(res.body.message).to.equal('Invalid user details');
+        done();
+      });
+  });
+  it('should not parseInt if Id has alphabet', (done) => {
```
Comment (iAmao) [inline test/users.test.spec.js]: This test is testing for the same case as above. We don't need it

File: server/controllers/userController.js [hunk 12]
```diff
@@ -300,6 +303,86 @@ const userController = {
         return res.status(200).jsend.success({ followed, follower });
       })
       .catch(() => res.status(500).jsend.error({ message: 'we could not get you details at this time . Please try again' }));
+  },
+  /**
+   * @method updateProfile
+   * @description Allows users to edit their profile
+   * @param {Object} req The request object
+   * @param {Object} res The the response object
```
Comment (iAmao) [inline server/controllers/userController.js]: take out one of the `the`

File: server/controllers/userController.js [hunk 13]
```diff
@@ -300,6 +303,86 @@ const userController = {
...
+   * @returns {Object} The modified profile details
+   */
+  updateProfile: (req, res) => {
+    multerUploads(req, res, () => {
+      const { id } = req.currentUser;
+      const {
+        firstname, lastname, username, email, bio,
+      } = req.body;
+      const interest = (req.body.interest) ? req.body.interest.split(',') : null;
```
Comment (iAmao) [inline server/controllers/userController.js]: The braces are not needed here though

File: server/controllers/userController.js [hunk 14]
```diff
@@ -300,6 +303,86 @@ const userController = {
...
+   * @method getUserProfile
+   * @description Allows users to view other user's profile
+   * @param {Object} req The request object
+   * @param {Object} res The the response object
+   * @returns {Object} The modified profile details
+   */
+  getUserProfile: (req, res) => {
+    const userId = parsedId(req.params.userId);
+    if (!(Number.isInteger(userId))) {
+      return res.status(400).jsend.error({
```
Comment (iAmao) [inline server/controllers/userController.js]: Lets just return 404 for this case

**Review by iAmao — CHANGES_REQUESTED** [review]
Nice 👍  but, there were some style violations you need to address. Also, some suggestions.

**Review by iAmao — APPROVED** [review]

**Pull request merged**